# K Nearest Neighbor

> Q: Write a program to implement k-Nearest Neighbour algorithm to classify the iris data set. Print both correct and wrong predictions.

### Reference to understand KNN
1. Medium Article: https://medium.com/@srishtisawla/k-nearest-neighbors-f77f6ee6b7f5
2. KNN Explaination in Youtube : https://www.youtube.com/watch?v=6kZ-OPLNcgE

## Aim :
Our aim to find which class does the Flower Feature belongs to.

> __1.Dataset__ - Iris dataset is a popular dataset which contains the feature about the flower interms of Image Tensor, Class Label etc.

> __2.Train-Test-Split__ - sklearn also provides a library from which we can divide the whole dataset into certain amount of Train & Test data. Which helps us to train the model as well as test it over it's learning

> __3. pandas & numpy__ - you guys might have probabily used but still, pandas used to handle the data in the form of dataframe and numpy for extra matrix, vector operation with ease. 

> __4. Model__ - KNN model can be easily implemented by ```sklearn.models``` library.

In [36]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_iris
import numpy as np
import pandas as pd

### 1. Data Preprocessing

In [37]:
# Class to get iris data
iris_data = load_iris()

# Train test split using the object of load_iris class
x_train, x_test, y_train, y_test = train_test_split(iris_data.data, iris_data.target)

print('Shape of Train-x:',x_train.shape, 'Shape of Test-x:',x_test.shape)
print('Shape of Train-y:',y_train.shape, 'Shape of Test-y:',y_test.shape)

Shape of Train-x: (112, 4) Shape of Test-x: (38, 4)
Shape of Train-y: (112,) Shape of Test-y: (38,)


Here shape is 2D value Means that, there are 112 rows of tuples of 4 Attributes are present.

3 Type of flowers are present: ['setosa' 'versicolor' 'virginica'] <-- Can be obtained by ```print(iris_data.target_names)```

### 2. Model Preparation & training:
Model is trained by __train__ dataset provided, with neighbors __1__.

In [38]:
# Create object of KNN classifier class
knn = KNeighborsClassifier(n_neighbors=1)

# Training
knn.fit(x_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=1, p=2,
           weights='uniform')

### 3. Predict using ```test``` data

In [39]:
y_pred = knn.predict(x_test)

for i in range(10):
    print('> Actual ',iris_data.target_names[y_train[i]],' -- Predicted ',iris_data.target_names[y_test[i]])          

> Actual  versicolor  -- Predicted  setosa
> Actual  versicolor  -- Predicted  virginica
> Actual  versicolor  -- Predicted  versicolor
> Actual  virginica  -- Predicted  versicolor
> Actual  setosa  -- Predicted  setosa
> Actual  versicolor  -- Predicted  setosa
> Actual  virginica  -- Predicted  versicolor
> Actual  setosa  -- Predicted  virginica
> Actual  setosa  -- Predicted  setosa
> Actual  versicolor  -- Predicted  virginica


In [40]:
print('Accuracy of KNN :', knn.score(x_test, y_test))

Accuracy of KNN : 0.9473684210526315
